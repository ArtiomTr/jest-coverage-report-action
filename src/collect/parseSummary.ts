import {
    coveredBranchesCounter,
    coveredLinesCounter,
    standardCoveredCounter,
    standardTotalCounter,
    totalBranchesCounter,
    totalLinesCounter,
} from './counters';
import { getSummary } from './getSummary';
import { JsonReport } from '../typings/JsonReport';

export const parseSummary = (jsonReport: JsonReport) => {
    return [
        getSummary(
            jsonReport.coverageMap,
            standardTotalCounter('s'),
            standardCoveredCounter('s'),
            'Statements'
        ),
        getSummary(
            jsonReport.coverageMap,
            totalBranchesCounter,
            coveredBranchesCounter,
            'Branches'
        ),
        getSummary(
            jsonReport.coverageMap,
            standardTotalCounter('f'),
            standardCoveredCounter('f'),
            'Functions'
        ),
        getSummary(
            jsonReport.coverageMap,
            totalLinesCounter,
            coveredLinesCounter,
            'Lines'
        ),
    ];
};
