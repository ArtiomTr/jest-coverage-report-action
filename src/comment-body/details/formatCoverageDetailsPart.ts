import markdownTable from 'markdown-table';

import { getFileCoverageDetailRow } from './getFileCoverageDetailRow';
import { ParsedCoverageDetails } from '../../collect-coverage/parseCoverageDetails';
import { details } from '../strings.json';
import { createMarkdownSpoiler } from '../utils/createMarkdownSpoiler';
import { formatHeadingAndTable } from '../utils/formatHeadingAndTable';

export type DetailsFormatOptions = {
    summary: string;
    heading: string;
};

export const formatCoverageDetailsPart = (
    formatOptions: DetailsFormatOptions,
    headDetails: ParsedCoverageDetails,
    baseDetails?: ParsedCoverageDetails
): string | undefined => {
    const { summary, heading } = formatOptions;

    const tableContent = Object.keys(headDetails).map((filename) =>
        getFileCoverageDetailRow(
            filename,
            headDetails[filename],
            baseDetails?.[filename]
        )
    );

    if (tableContent.length > 0) {
        return createMarkdownSpoiler({
            body: formatHeadingAndTable(
                heading,
                markdownTable([details.columnHeaders, ...tableContent])
            ),
            summary,
        });
    }

    return undefined;
};
