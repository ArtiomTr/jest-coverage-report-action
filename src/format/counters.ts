import { FileCoverage } from '../typings/JsonReport';

export const standardTotalCounter = (key: keyof FileCoverage) => (
    value: FileCoverage
) => Object.values(value[key]).length;

export const standardCoveredCounter = (key: keyof FileCoverage) => (
    value: FileCoverage
) => Object.values(value[key]).filter((hits) => hits > 0).length;

export const totalBranchesCounter = (value: FileCoverage) =>
    Object.values(value.b).reduce((acc, branch) => acc + branch.length, 0);

export const coveredBranchesCounter = (value: FileCoverage) =>
    Object.values(value.b).reduce(
        (acc, branch) => acc + branch.filter((hits) => hits > 0).length,
        0
    );

export const totalLinesCounter = (value: FileCoverage) => {
    const stats: Record<string, number> = getLineCoverage(value);
    return Object.keys(stats).length;
};

export const coveredLinesCounter = (value: FileCoverage) => {
    const stats: Record<string, number> = getLineCoverage(value);
    return Object.values(stats).filter((v) => !!v).length;
};

const getLineCoverage = (value: FileCoverage) => {
    const statementMap = value.statementMap;
    const statements = value.s;

    return Object.entries(statements).reduce((acc, [st, count]) => {
        const _st: number = parseInt(st);

        if (!statementMap[_st]) {
            return acc;
        }
        const { line } = statementMap[_st].start;
        const prevVal = acc[line];
        if (prevVal === undefined || prevVal < count) {
            acc[line] = count;
        }
        return acc;
    }, {} as Record<string, number>);
};
