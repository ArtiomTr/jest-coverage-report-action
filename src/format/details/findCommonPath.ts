import { dirname } from 'path';

/**
 * Find a common path from a list of filepaths.
 */
export const findCommonPath = (filepaths: string[]): string => {
    let commonRoot = '';
    if (filepaths.length) {
        // If the paths are sorted, any prefix common to all paths will be common to the sorted first and last strings.
        const sortedPaths = [...filepaths].sort();
        const first = sortedPaths[0];
        const last = sortedPaths[sortedPaths.length - 1];

        const len = Math.min(first.length, last.length);
        for (let i = 0; i < len; i++) {
            const ch1 = first[i];
            const ch2 = last[i];
            if (ch1 == ch2) {
                commonRoot += ch1;
            } else {
                break;
            }
        }

        // If it doesn't appear to be a directory (partial filename), get the parent directory
        if (commonRoot.length && commonRoot[commonRoot.length - 1] !== '/') {
            commonRoot = dirname(commonRoot) + '/';
        }
    }
    return commonRoot;
};
