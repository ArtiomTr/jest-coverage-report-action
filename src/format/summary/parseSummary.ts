import { getSummary } from './getSummary';
import { JsonReport } from '../../typings/JsonReport';
import { i18n } from '../../utils/i18n';
import {
    coveredBranchesCounter,
    coveredLinesCounter,
    standardCoveredCounter,
    standardTotalCounter,
    totalBranchesCounter,
    totalLinesCounter,
} from '../counters';

export const parseSummary = (jsonReport: JsonReport) => {
    return [
        getSummary(
            jsonReport.coverageMap,
            standardTotalCounter('s'),
            standardCoveredCounter('s'),
            i18n('statements')
        ),
        getSummary(
            jsonReport.coverageMap,
            totalBranchesCounter,
            coveredBranchesCounter,
            i18n('branches')
        ),
        getSummary(
            jsonReport.coverageMap,
            standardTotalCounter('f'),
            standardCoveredCounter('f'),
            i18n('functions')
        ),
        getSummary(
            jsonReport.coverageMap,
            totalLinesCounter,
            coveredLinesCounter,
            i18n('lines')
        ),
    ];
};
