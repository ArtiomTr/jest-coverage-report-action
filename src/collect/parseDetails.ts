import {
    coveredBranchesCounter,
    coveredLinesCounter,
    standardCoveredCounter,
    standardTotalCounter,
    totalBranchesCounter,
    totalLinesCounter,
} from './counters';
import { findCommonPath } from './findCommonPath';
import { getPercents } from './getPercents';
import { CoverageDetailsMap } from '../typings/Coverage';
import { JsonReport } from '../typings/JsonReport';

export const parseDetails = (jsonReport: JsonReport) => {
    // Find common root directory
    let trimPath = 0;
    const filepaths = Object.keys(jsonReport.coverageMap);
    if (filepaths.length) {
        const commonRootPath = findCommonPath(filepaths);
        trimPath = commonRootPath.length;
    }

    return Object.entries(jsonReport.coverageMap).reduce<CoverageDetailsMap>(
        (acc, [filename, fileCoverage]) => {
            const normalizedFilename = filename.substr(trimPath);
            acc[normalizedFilename] = {
                filename: normalizedFilename,
                statements: getPercents(
                    standardCoveredCounter('s')(fileCoverage),
                    standardTotalCounter('s')(fileCoverage)
                ),
                functions: getPercents(
                    standardCoveredCounter('f')(fileCoverage),
                    standardTotalCounter('f')(fileCoverage)
                ),
                branches: getPercents(
                    coveredBranchesCounter(fileCoverage),
                    totalBranchesCounter(fileCoverage)
                ),
                lines: getPercents(
                    coveredLinesCounter(fileCoverage),
                    totalLinesCounter(fileCoverage)
                ),
            };
            return acc;
        },
        {}
    );
};
