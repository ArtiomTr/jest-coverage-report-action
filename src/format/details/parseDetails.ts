import { findCommonPath } from './findCommonPath';
import { CoverageDetailsMap } from '../../typings/Coverage';
import { JsonReport } from '../../typings/JsonReport';
import { getFileCoverageMap } from '../../utils/getFileCoverageMap';
import { getPercents } from '../getPercents';

export const parseDetails = (jsonReport: JsonReport) => {
    // Find common root directory
    const filepaths = Object.keys(jsonReport.coverageMap);
    const trimPath = findCommonPath(filepaths).length;

    const fileCoverageMap = getFileCoverageMap(jsonReport);

    return Object.entries(fileCoverageMap).reduce<CoverageDetailsMap>(
        (acc, [filename, coverage]) => {
            const normalizedFilename = filename.substring(trimPath);

            acc[normalizedFilename] = {
                filename: normalizedFilename,
                statements: getPercents(
                    coverage.coveredStatements,
                    coverage.totalStatements
                ),
                branches: getPercents(
                    coverage.coveredBranches,
                    coverage.totalBranches
                ),
                functions: getPercents(
                    coverage.coveredFunctions,
                    coverage.totalFunctions
                ),
                lines: getPercents(coverage.coveredLines, coverage.totalLines),
            };

            return acc;
        },
        {}
    );
};
