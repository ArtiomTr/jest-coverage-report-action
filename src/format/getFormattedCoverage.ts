import { formatCoverageDetails } from './details/formatCoverageDetails';
import { formatCoverageSummary } from './summary/formatCoverageSummary';
import { CoverageDetailsMap, CoverageSummary } from '../typings/Coverage';
import { i18n } from '../utils/i18n';

export const getFormattedCoverage = (
    headSummary: Array<CoverageSummary>,
    baseSummary: Array<CoverageSummary> | undefined,
    headDetails: CoverageDetailsMap,
    baseDetails: CoverageDetailsMap | undefined,
    threshold: number | undefined,
    hideDetails: boolean | undefined
): string =>
    [
        formatCoverageSummary(headSummary, baseSummary, threshold),
        !hideDetails
            ? formatCoverageDetails(headDetails, baseDetails, threshold)
            : `> ${i18n('detailsHidden')}`,
    ]
        .filter(Boolean)
        .join('\n');
