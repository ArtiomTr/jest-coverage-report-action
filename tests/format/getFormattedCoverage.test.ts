import { parseDetails } from '../../src/format/details/parseDetails';
import { getFormattedCoverage } from '../../src/format/getFormattedCoverage';
import { parseSummary } from '../../src/format/summary/parseSummary';
import jsonReport from '../mock-data/jsonReport.json';

describe('getFormattedCoverage', () => {
    it('should match snapshots', () => {
        expect(
            getFormattedCoverage(
                parseSummary(jsonReport),
                undefined,
                parseDetails(jsonReport),
                undefined,
                undefined,
                false
            )
        ).toMatchSnapshot();
    });
});
