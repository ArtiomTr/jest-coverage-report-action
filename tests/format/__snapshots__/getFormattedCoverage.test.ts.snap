// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`getFormattedCoverage > should match snapshots 1`] = `
"| <div title=\\"Status of coverage:&#10; 游릭 - ok&#10; 游리 - slightly more than threshold&#10; 游댮 - under the threshold\\">St.<sup>:grey_question:</sup></div> | Category   | Percentage | Covered / Total |
| :----------------------------------------------------------------------------------------------------------------------------------------------------: | :--------- | :--------- | :-------------: |
|                                                                           游릭                                                                           | Statements | 81.82%     |      27/33      |
|                                                                           游릭                                                                           | Branches   | 100%       |       8/8       |
|                                                                           游리                                                                           | Functions  | 63.64%     |       7/11      |
|                                                                           游릭                                                                           | Lines      | 80.65%     |      25/31      |

"
`;
